oauth2/src/main/java/com/linecorp/armeria/client/auth/oauth2/AbstractOAuth2AuthorizationGrant.java

```diff
@@ -118,6 +99,84 @@ private static boolean shallObtainInsteadOfUpdate(Throwable throwable) {
      */
     @Override
     public final CompletionStage<GrantedOAuth2AccessToken> getAccessToken() {
-        return tokenManager.get();
+        final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
+        final CompletableFuture<GrantedOAuth2AccessToken> tokenFuture = this.tokenFuture;
+
+        tokenFuture.handle((token, ignored) -> {
+            if (isValidToken(token)) {
+                future.complete(token);
+                return null;
+            }
+
+            final Supplier<CompletionStage<GrantedOAuth2AccessToken>> tokenIssuingFunc;
+            if (tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
+                tokenIssuingFunc = token != null && token.isRefreshable() ? () -> refreshAccessToken(token)
+                                                                          : () -> issueAccessToken(token);
+            } else {
+                tokenIssuingFunc = () -> this.tokenFuture;
+            }
+
+            final CompletionStage<GrantedOAuth2AccessToken> tokenIssuingFuture = RequestContext.mapCurrent(
+                            ctx -> ctx.makeContextAware(tokenIssuingFunc.get()), tokenIssuingFunc);
+            tokenIssuingFuture.handle((newToken, cause) -> {
+                if (cause != null) {
+                    future.completeExceptionally(cause);
+                } else {
+                    future.complete(newToken);
+                    if (saveTokenFunc != null) {
+                        // suppress 'ReturnValueIgnored' warning
+                        final CompletableFuture<Void> unused = saveTokenFunc.apply(newToken);
+                    }
+                }
+                return null;
+            });
+            return null;
+        });
+        return future;
+    }
+
+    private boolean isValidToken(@Nullable GrantedOAuth2AccessToken token) {
+        return token != null && token.isValid(Instant.now().plus(refreshBefore));
+    }
+
+    private CompletableFuture<GrantedOAuth2AccessToken> issueAccessToken(
+            @Nullable GrantedOAuth2AccessToken token) {
+        if (token == null && loadTokenFunc != null) {
+            return loadTokenFunc.get().thenCompose(storedToken -> {
+                if (storedToken.isValid()) {
+                    return UnmodifiableFuture.completedFuture(storedToken);
+                }
+                return obtainAccessToken(null);
+            });
+        }
+        return obtainAccessToken(token);
+    }
+
+    private CompletableFuture<GrantedOAuth2AccessToken> refreshAccessToken(GrantedOAuth2AccessToken token) {
+        final CompletableFuture<GrantedOAuth2AccessToken> future = new CompletableFuture<>();
```

Comment:
How about passing the `future` https://github.com/line/armeria/pull/3618/files#diff-866e837e1f45041da819b2fed992548e97c9b2a8a72071e68c374f9d4678e53aR123 instead of creating a new one?
Then, we can change here https://github.com/line/armeria/pull/3618/files#diff-866e837e1f45041da819b2fed992548e97c9b2a8a72071e68c374f9d4678e53aR123

```java
if (tokenFutureUpdater.compareAndSet(this, tokenFuture, future)) {
    if (token != null && token.isRefreshable()) {
        refreshAccessToken(token, future);
        return null;
    }
    ...
}
```

Reply:
Nice, let me update the PR and ping you again!

Reply:
Done!